packages/http-server-javascript/src/common/reference.ts

```diff
@@ -0,0 +1,296 @@
+      }
+
+      // TODO/witemple: I believe this is going to declare all template instances in the module of the template itself, which
+      // might not be desirable.
+      const parentModule = createOrGetModuleForNamespace(ctx, effectiveModel.namespace);
+
+      module.imports.push({
+        binder: [templatedName.pascalCase],
+        from: parentModule,
+      });
+
+      return templatedName.pascalCase;
+    }
+    case "Union": {
+      if (type.variants.size === 0) return "never";
+      else if (type.variants.size === 1)
+        return emitTypeReference(ctx, [...type.variants.values()][0], position, module, options);
+
+      if (options.requireDeclaration) {
+        if (type.name) {
+          const nameCase = parseCase(type.name);
+
+          ctx.typeQueue.add(type);
+
+          module.imports.push({
+            binder: [nameCase.pascalCase],
+            from: createOrGetModuleForNamespace(ctx, type.namespace!),
+          });
+
+          return type.name;
+        } else {
+          const existingSyntheticName = ctx.syntheticNames.get(type);
+
+          if (existingSyntheticName) {
+            module.imports.push({
+              binder: [existingSyntheticName],
+              from: ctx.syntheticModule,
+            });
+
+            return existingSyntheticName;
+          } else {
+            const altName = options.altName;
+
+            if (!altName) {
+              throw new Error("UNREACHABLE: anonymous union without preferredAlternativeName");
+            }
+
+            ctx.synthetics.push({
+              kind: "anonymous",
+              name: altName,
+              underlying: type,
+            });
+
+            module.imports.push({
+              binder: [altName],
+              from: ctx.syntheticModule,
+            });
+
+            ctx.syntheticNames.set(type, altName);
+
+            return altName;
+          }
+        }
+      } else {
+        return emitUnionType(ctx, [...type.variants.values()], module);
+      }
+    }
+    case "Enum": {
+      ctx.typeQueue.add(type);
+
+      const name = parseCase(type.name).pascalCase;
+
+      module.imports.push({
+        binder: [name],
+        from: createOrGetModuleForNamespace(ctx, type.namespace!),
+      });
+
+      return name;
+    }
+    case "String":
+      return JSON.stringify(type.value);
+    case "Number":
+    case "Boolean":
+      return String(type.value);
+    case "Intrinsic":
+      switch (type.name) {
+        case "never":
+          return "never";
+        case "null":
+          return "null";
+        case "void":
+          // It's a bit strange to have a void property, but it's possible, and TypeScript allows it. Void is simply
+          // only assignable from undefined or void itself.
+          return "void";
+        case "ErrorType":
+          throw new Error("UNREACHABLE: Encountered ErrorType in emitTypeReference");
```

Comment:
i recommend using `compilerAssert` here instead you can also pass a 3rd arg which is a diagnostic target in case this actually get hit gives a bit more help

Reply:
`compilerAssert` does not allow `typeof NoTarget`, which is probably it's own problem I'll make an issue for. I'll work around that here for now by only providing `position` if the value is not NoTarget, otherwise passing `type`, which will just be ErrorType.

Reply:
It also seems to interact strangely with control flow analysis. The `compilerAssert` function isn't analyzed as a break, but putting a `break` after the `compilerAssert` yields an unreachable code warning.

Reply:
You can then technically not pass the 3rd arg if you don’t have a target

Reply:
Yeah so I updated this to pass `position` if there is a target or the type itself otherwise. I also just added a defensive return afterwords since TypeScript isn't correctly analyzing the control flow.